Review followup: PR #24592 — GH#24591: fix: record core routine run metrics

## Unaddressed review bot suggestions

PR #24592 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/24592

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/pulse-routines.sh:128`
- `.agents/scripts/pulse-routines.sh:185`
- `.agents/scripts/setup/modules/schedulers-platform.sh:515`
- `.agents/scripts/setup/modules/schedulers-platform.sh:529`
- `.agents/scripts/setup/modules/schedulers.sh:499`
- `.agents/scripts/tests/test-launchd-install-if-changed.sh:810`
- `.agents/scripts/tests/test-routine-tracking-updates.sh:173`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/setup/modules/schedulers-platform.sh:515`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328047)

```diff
@@ -507,12 +507,27 @@ _run_profile_readme_init() {
 	return 0
 }
 
+_core_routine_logged_command() {
+	local routine_id="$1"
+	local exec_command="$2"
+	local log_helper="\$HOME/.aidevops/agents/scripts/routine-log-helper.sh"
+	# shellcheck disable=SC2016 # command string is expanded by the generated scheduler shell.
+	printf 'start_epoch=$(date +%%s); status=success; %s; rc=$?; end_epoch=$(date +%%s); duration=$((end_epoch - start_epoch)); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "%s" ]; then "%s" update "%s" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"' \
```

> ![security-high](https://www.gstatic.com/codereviewagent/security-high-priority.svg) ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> To prevent command injection from shell metacharacters in crontab entries, wrap variables in single quotes. This is a POSIX-compliant method to prevent all shell expansion, including `$(...)`, backticks, and variable expansion. Additionally, use default values in the arithmetic expansion to prevent syntax errors if `start_epoch` or `end_epoch` is empty.
> 
> ```suggestion
> 	printf 'start_epoch=$(date +%%s); status=success; %s; rc=$?; end_epoch=$(date +%%s); duration=$(( ${end_epoch:-0} - ${start_epoch:-0} )); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "%s" ]; then "%s" update "%s" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"' \
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. To prevent command injection from shell metacharacters in crontab entries, wrap variables in single quotes to prevent all shell expansion.
> 2. In shell scripts with set -e, arithmetic expansion can fail if variables are empty. Use parameter expansion with a default value to handle empty strings robustly.
> </details>

#### gemini-code-assist on `.agents/scripts/setup/modules/schedulers-platform.sh:529`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328051)

```diff
+		"$routine_id"
+	return 0
+}
+
 _install_profile_readme_launchd() {
 	local pr_label="$1"
 	local pr_script="$2"
 	local pr_plist="$HOME/Library/LaunchAgents/${pr_label}.plist"
 	local _xml_pr_script _xml_pr_home
-	_xml_pr_script=$(_xml_escape "$pr_script")
+	local pr_command
+	pr_command=$(_core_routine_logged_command "r908" "\"${pr_script}\" update")
```

> ![security-high](https://www.gstatic.com/codereviewagent/security-high-priority.svg) ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> Wrap the script path variable in single quotes instead of double quotes to prevent command injection from shell metacharacters in crontab/scheduler entries.
> 
> ```suggestion
> 	pr_command=$(_core_routine_logged_command "r908" "'${pr_script}' update")
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. To prevent command injection from shell metacharacters in crontab entries, wrap variables in single quotes to prevent all shell expansion.
> </details>

#### gemini-code-assist on `.agents/scripts/setup/modules/schedulers.sh:499`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328053)

```diff
@@ -495,7 +495,9 @@ setup_screen_time_snapshot() {
 
 		# XML-escape paths for safe plist embedding
 		local _xml_st_script _xml_st_home
-		_xml_st_script=$(_xml_escape "$st_script")
+		local st_command
+		st_command=$(_core_routine_logged_command "r909" "\"${st_script}\" snapshot")
```

> ![security-high](https://www.gstatic.com/codereviewagent/security-high-priority.svg) ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> Wrap the script path variable in single quotes instead of double quotes to prevent command injection from shell metacharacters in crontab/scheduler entries.
> 
> ```suggestion
> 		st_command=$(_core_routine_logged_command "r909" "'${st_script}' snapshot")
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. To prevent command injection from shell metacharacters in crontab entries, wrap variables in single quotes to prevent all shell expansion.
> </details>

#### gemini-code-assist on `.agents/scripts/pulse-routines.sh:128`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328054)

```diff
@@ -118,6 +120,12 @@ _routine_execute() {
 		if [[ ! -x "$script_path" ]]; then
 			echo "[pulse-wrapper] routine ${routine_id}: script not found or not executable: ${script_path}" >>"$LOGFILE"
 			_routine_update_state "$routine_id" "failure"
+			if [[ -x "$ROUTINE_LOG_HELPER" ]]; then
+				local ended_epoch
+				ended_epoch=$(date +%s)
+				local duration=$((ended_epoch - started_epoch))
+				"$ROUTINE_LOG_HELPER" update "$routine_id" --status failure --duration "$duration" 2>/dev/null || true
+			fi
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> To prevent potential shell syntax errors during arithmetic expansion if `started_epoch` or `ended_epoch` is empty or unset, use default values (e.g., `${ended_epoch:-0}` and `${started_epoch:-0}`).
> 
> ```suggestion
> 			if [[ -x "$ROUTINE_LOG_HELPER" ]]; then
> 				local ended_epoch
> 				ended_epoch=$(date +%s)
> 				local duration=$(( ${ended_epoch:-0} - ${started_epoch:-0} ))
> 				"$ROUTINE_LOG_HELPER" update "$routine_id" --status failure --duration "$duration" 2>/dev/null || true
> 			fi
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts with set -e, arithmetic expansion can fail if variables are empty. Use parameter expansion with a default value to handle empty strings robustly.
> </details>

#### gemini-code-assist on `.agents/scripts/pulse-routines.sh:185`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328057)

```diff
@@ -171,7 +179,10 @@ _routine_execute() {
 
 	# Call routine-log-helper.sh if available (t1926)
 	if [[ -x "$ROUTINE_LOG_HELPER" ]]; then
-		"$ROUTINE_LOG_HELPER" update "$routine_id" "$status" 2>/dev/null || true
+		local ended_epoch
+		ended_epoch=$(date +%s)
+		local duration=$((ended_epoch - started_epoch))
+		"$ROUTINE_LOG_HELPER" update "$routine_id" --status "$status" --duration "$duration" 2>/dev/null || true
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> To prevent potential shell syntax errors during arithmetic expansion if `started_epoch` or `ended_epoch` is empty or unset, use default values (e.g., `${ended_epoch:-0}` and `${started_epoch:-0}`).
> 
> ```suggestion
> 		local ended_epoch
> 		ended_epoch=$(date +%s)
> 		local duration=$(( ${ended_epoch:-0} - ${started_epoch:-0} ))
> 		"$ROUTINE_LOG_HELPER" update "$routine_id" --status "$status" --duration "$duration" 2>/dev/null || true
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts with set -e, arithmetic expansion can fail if variables are empty. Use parameter expansion with a default value to handle empty strings robustly.
> </details>

#### gemini-code-assist on `.agents/scripts/tests/test-routine-tracking-updates.sh:173`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328060)

```diff
+	# shellcheck source=/dev/null
+	source "$PULSE_ROUTINES_SH"
+
+	_routine_execute "r777" "sample" "scripts/sample.sh" "" "$TEST_DIR"
+
+	local args
+	args=$(<"$ROUTINE_LOG_CAPTURE")
+	if [[ "$args" == update\ r777\ --status\ success\ --duration\ * ]]; then
+		print_result "pulse routine update passes flags and duration" 0
+	else
+		print_result "pulse routine update passes flags and duration" 1 "$args"
+	fi
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> If `_routine_execute` fails or doesn't call the helper, `ROUTINE_LOG_CAPTURE` won't be created. Reading from a non-existent file using `$(<...)` will cause the test script to crash abruptly due to `set -e`. Check if the file exists first to allow the test to fail gracefully with a descriptive message.
> 
> ```suggestion
> 	local args=""
> 	if [[ -f "$ROUTINE_LOG_CAPTURE" ]]; then
> 		args=$(<"$ROUTINE_LOG_CAPTURE")
> 	fi
> 	if [[ "$args" == update\ r777\ --status\ success\ --duration\ * ]]; then
> 		print_result "pulse routine update passes flags and duration" 0
> 	else
> 		print_result "pulse routine update passes flags and duration" 1 "$args"
> 	fi
> ```

#### gemini-code-assist on `.agents/scripts/tests/test-launchd-install-if-changed.sh:810`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/24592#discussion_r3377328065)

```diff
@@ -806,21 +806,21 @@ test_profile_readme_install_does_not_kickstart() {
 	<key>ProgramArguments</key>
 	<array>
 		<string>/bin/bash</string>
-		<string>/fake/profile-readme-helper.sh</string>
-		<string>update</string>
+		<string>-lc</string>
+		<string>start_epoch=$(date +%s); status=success; "/fake/profile-readme-helper.sh" update; rc=$?; end_epoch=$(date +%s); duration=$((end_epoch - start_epoch)); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" ]; then "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" update "r908" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"</string>
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> Update the expected plist content in the test to match the single-quoted script path and robust arithmetic expansion implemented in the main scheduler code.
> 
> ```suggestion
> 			<string>start_epoch=$(date +%s); status=success; '/fake/profile-readme-helper.sh' update; rc=$?; end_epoch=$(date +%s); duration=$(( ${end_epoch:-0} - ${start_epoch:-0} )); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" ]; then "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" update "r908" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"</string>
> ```

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/24592#pullrequestreview-4454590988)

> ## Code Review
> 
> This pull request introduces routine execution duration tracking and status logging across scheduler platforms and pulse routines. It adds a helper function `_core_routine_logged_command` to wrap scheduled commands with execution metrics, updates launchd and systemd schedulers to use this wrapper, and adds corresponding tests. The review feedback suggests enhancing security and robustness by wrapping script paths in single quotes to prevent command injection, using default values in arithmetic expansions to handle empty variables safely, checking for file existence in tests to prevent crashes under `set -e`, and updating test expectations accordingly.
> 
> > [!IMPORTANT]
> > The [consumer version of Gemini Code Assist on GitHub](https://developers.google.com/gemini-code-assist/docs/review-repo-code) is being sunset. Starting **June 18, 2026**, new organization installations will be blocked, and all code review activity will officially cease on **July 17, 2026**.
> > For more details on the timeline and next steps, please review the [Help Documentation](https://developers.google.com/gemini-code-assist/docs/deprecations/consumer-code-review).
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.20.41 automated scan.

--- a/.agents/scripts/pulse-routines.sh
+++ b/.agents/scripts/pulse-routines.sh
@@ -123,7 +123,7 @@ _routine_execute() {
 			if [[ -x "$ROUTINE_LOG_HELPER" ]]; then
 				local ended_epoch
 				ended_epoch=$(date +%s)
-				local duration=$((ended_epoch - started_epoch))
+				local duration=$(( ${ended_epoch:-0} - ${started_epoch:-0} ))
 				"$ROUTINE_LOG_HELPER" update "$routine_id" --status failure --duration "$duration" 2>/dev/null || true
 			fi
 			return 1
@@ -181,7 +181,7 @@ _routine_execute() {
 	if [[ -x "$ROUTINE_LOG_HELPER" ]]; then
 		local ended_epoch
 		ended_epoch=$(date +%s)
-		local duration=$((ended_epoch - started_epoch))
+		local duration=$(( ${ended_epoch:-0} - ${started_epoch:-0} ))
 		"$ROUTINE_LOG_HELPER" update "$routine_id" --status "$status" --duration "$duration" 2>/dev/null || true
 	fi
 

--- a/.agents/scripts/setup/modules/schedulers-platform.sh
+++ b/.agents/scripts/setup/modules/schedulers-platform.sh
@@ -507,12 +507,20 @@ _run_profile_readme_init() {
 	return 0
 }
 
+_core_routine_shell_quote() {
+	local value="$1"
+	printf "'"
+	printf '%s' "$value" | sed "s/'/'\\\\''/g"
+	printf "'"
+	return 0
+}
+
 _core_routine_logged_command() {
 	local routine_id="$1"
 	local exec_command="$2"
 	local log_helper="\$HOME/.aidevops/agents/scripts/routine-log-helper.sh"
 	# shellcheck disable=SC2016 # command string is expanded by the generated scheduler shell.
-	printf 'start_epoch=$(date +%%s); status=success; %s; rc=$?; end_epoch=$(date +%%s); duration=$((end_epoch - start_epoch)); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "%s" ]; then "%s" update "%s" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"' \
+	printf 'start_epoch=$(date +%%s); status=success; %s; rc=$?; end_epoch=$(date +%%s); duration=$(( ${end_epoch:-0} - ${start_epoch:-0} )); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "%s" ]; then "%s" update "%s" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"' \
 		"$exec_command" \
 		"$log_helper" \
 		"$log_helper" \
@@ -526,7 +534,7 @@ _install_profile_readme_launchd() {
 	local pr_plist="$HOME/Library/LaunchAgents/${pr_label}.plist"
 	local _xml_pr_script _xml_pr_home
 	local pr_command
-	pr_command=$(_core_routine_logged_command "r908" "\"${pr_script}\" update")
+	pr_command=$(_core_routine_logged_command "r908" "$(_core_routine_shell_quote "$pr_script") update")
 	_xml_pr_script=$(_xml_escape "$pr_command")
 	_xml_pr_home=$(_xml_escape "$HOME")
 

--- a/.agents/scripts/setup/modules/schedulers.sh
+++ b/.agents/scripts/setup/modules/schedulers.sh
@@ -496,7 +496,7 @@ setup_screen_time_snapshot() {
 		# XML-escape paths for safe plist embedding
 		local _xml_st_script _xml_st_home
 		local st_command
-		st_command=$(_core_routine_logged_command "r909" "\"${st_script}\" snapshot")
+		st_command=$(_core_routine_logged_command "r909" "$(_core_routine_shell_quote "$st_script") snapshot")
 		_xml_st_script=$(_xml_escape "$st_command")
 		_xml_st_home=$(_xml_escape "$HOME")
 

--- a/.agents/scripts/tests/test-launchd-install-if-changed.sh
+++ b/.agents/scripts/tests/test-launchd-install-if-changed.sh
@@ -807,7 +807,7 @@ test_profile_readme_install_does_not_kickstart() {
 	<array>
 		<string>/bin/bash</string>
 		<string>-lc</string>
-		<string>start_epoch=$(date +%s); status=success; "/fake/profile-readme-helper.sh" update; rc=$?; end_epoch=$(date +%s); duration=$((end_epoch - start_epoch)); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" ]; then "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" update "r908" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"</string>
+		<string>start_epoch=$(date +%s); status=success; '/fake/profile-readme-helper.sh' update; rc=$?; end_epoch=$(date +%s); duration=$(( ${end_epoch:-0} - ${start_epoch:-0} )); if [ "$rc" -ne 0 ]; then status=failure; fi; if [ -x "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" ]; then "\$HOME/.aidevops/agents/scripts/routine-log-helper.sh" update "r908" --status "$status" --duration "$duration" >/dev/null 2>&1 || true; fi; exit "$rc"</string>
 	</array>
 	<key>StartInterval</key>
 	<integer>3600</integer>

--- a/.agents/scripts/tests/test-routine-tracking-updates.sh
+++ b/.agents/scripts/tests/test-routine-tracking-updates.sh
@@ -63,9 +63,14 @@ load_scheduler_helpers() {
 
 test_core_routine_logged_command_shape() {
 	local command_text
-	command_text=$(_core_routine_logged_command "r908" '"/tmp/profile-readme-helper.sh" update')
-	if [[ "$command_text" != *'"/tmp/profile-readme-helper.sh" update'* ]]; then
+	command_text=$(_core_routine_logged_command "r908" "'$TEST_DIR/profile-readme-helper.sh' update")
+	if [[ "$command_text" != *"'$TEST_DIR/profile-readme-helper.sh' update"* ]]; then
 		print_result "core routine logged command keeps original command" 1 "$command_text"
+		return 0
+	fi
+	# shellcheck disable=SC2016 # the generated command must defer variable expansion to runtime.
+	if [[ "$command_text" != *'duration=$(( ${end_epoch:-0} - ${start_epoch:-0} ))'* ]]; then
+		print_result "core routine logged command uses robust duration arithmetic" 1 "$command_text"
 		return 0
 	fi
 	# shellcheck disable=SC2016 # the generated command must defer variable expansion to runtime.
@@ -164,8 +169,10 @@ HELPER
 
 	_routine_execute "r777" "sample" "scripts/sample.sh" "" "$TEST_DIR"
 
-	local args
-	args=$(<"$ROUTINE_LOG_CAPTURE")
+	local args=""
+	if [[ -f "$ROUTINE_LOG_CAPTURE" ]]; then
+		args=$(<"$ROUTINE_LOG_CAPTURE")
+	fi
 	if [[ "$args" == update\ r777\ --status\ success\ --duration\ * ]]; then
 		print_result "pulse routine update passes flags and duration" 0
 	else

